File: 2022/day-02.py

def score(i_play, outcome):
    return i_play + 1 + outcome * 3

def part_1_score(they_play, i_play):
    outcome = (i_play - they_play + 1) % 3
    return score(i_play, outcome)

input = open('input-02', 'r').read().strip()
guide = [(ord(line[0]) - ord('A'), ord(line[2]) - ord('X')) for line in input.split('\n')]

print(f'Part1: {sum([part_1_score(*round) for round in guide])}')


def part_2_score(they_play, outcome):
    i_play = (they_play + outcome - 1) % 3
    return score(i_play, outcome)

print(f'Part2: {sum([part_2_score(*round) for round in guide])}')

File: 2022/rust/src/days/day02.rs
fn score(i_play: u8, outcome: u8) -> u8 {
    return i_play + 1 + outcome * 3;
}

fn part_1_score(they_play: u8, i_play: u8) -> u8 {
    let outcome = (i_play + 4 - they_play) % 3;
    return score(i_play, outcome);
}


pub(crate) fn part1(input: &String) -> String {
    let buffer = input.as_bytes();
    let mut offset = 0;

    let mut total_score = 0;

    while offset < buffer.len() {
        total_score += part_1_score(
            buffer[offset] - 'A' as u8,
            buffer[offset + 2] - 'X' as u8
        ) as usize;
        offset += 4;
    }

    return total_score.to_string();
}


fn part_2_score(they_play: u8, outcome: u8) -> u8 {
    let i_play = (they_play + 2 + outcome) % 3;
    return score(i_play, outcome);
}

pub(crate) fn part2(input: &String) -> String {
    let buffer = input.as_bytes();
    let mut offset = 0;

    let mut total_score = 0;

    while offset < buffer.len() {
        total_score += part_2_score(
            buffer[offset] - 'A' as u8,
            buffer[offset + 2] - 'X' as u8
        ) as usize;
        offset += 4;
    }

    return total_score.to_string();
}